File: aula04 23_09/Python/Strategy.py
# Interface para a estratégia de fabricação de carros
class CarManufacturingStrategy:
    def manufacture(self):
        pass


# Implementações específicas de estratégias
class ElectricCarManufacturingStrategy(CarManufacturingStrategy):
    def manufacture(self):
        print("Fabricando um carro elétrico...")

class GasolineCarManufacturingStrategy(CarManufacturingStrategy):
    def manufacture(self):
        print("Fabricando um carro a gasolina...")


# Contexto da fábrica de carros
class CarFactory:
    def __init__(self, strategy):
        self.strategy = strategy

    def set_strategy(self, strategy):
        self.strategy = strategy

    def produce_car(self):
        self.strategy.manufacture()


# Exemplo de uso
if __name__ == "__main__":
    # Criação das estratégias de fabricação de carros
    electric_strategy = ElectricCarManufacturingStrategy()
    gasoline_strategy = GasolineCarManufacturingStrategy()

    # Criação da fábrica de carros com uma estratégia padrão
    factory = CarFactory(electric_strategy)

    # Produção de um carro usando a estratégia atual
    factory.produce_car()

    # Mudança da estratégia de fabricação em tempo de execução
    factory.set_strategy(gasoline_strategy)

    # Produção de outro carro usando a nova estratégia
    factory.produce_car()

File: aula04 23_09/Java/Strategy.java
// Interface para a estratégia de fabricação de carros
interface CarManufacturingStrategy {
    void manufacture();
}

// Implementações específicas de estratégias
class ElectricCarManufacturingStrategy implements CarManufacturingStrategy {
    @Override
    public void manufacture() {
        System.out.println("Fabricando um carro elétrico...");
    }
}

class GasolineCarManufacturingStrategy implements CarManufacturingStrategy {
    @Override
    public void manufacture() {
        System.out.println("Fabricando um carro a gasolina...");
    }
}

// Contexto da fábrica de carros
class CarFactory {
    private CarManufacturingStrategy strategy;

    public CarFactory(CarManufacturingStrategy strategy) {
        this.strategy = strategy;
    }

    public void setStrategy(CarManufacturingStrategy strategy) {
        this.strategy = strategy;
    }

    public void produceCar() {
        strategy.manufacture();
    }
}

// Exemplo de uso
public class Strategy {
    public static void main(String[] args) {
        // Criação das estratégias de fabricação de carros
        CarManufacturingStrategy electricStrategy = new ElectricCarManufacturingStrategy();
        CarManufacturingStrategy gasolineStrategy = new GasolineCarManufacturingStrategy();

        // Criação da fábrica de carros com uma estratégia padrão
        CarFactory factory = new CarFactory(electricStrategy);

        // Produção de um carro usando a estratégia atual
        factory.produceCar();

        // Mudança da estratégia de fabricação em tempo de execução
        factory.setStrategy(gasolineStrategy);

        // Produção de outro carro usando a nova estratégia
        factory.produceCar();
    }
}
